### ISSUE TITLE
Accept a Target as a value of "executable" argument to ctx.action

### ISSUE THREAD

**Issue by abergmeier** [bazelbuild/bazel#1447]
When trying to execute a `py_binary` on my Ubuntu 16.04

```
ctx.action(
    outputs = outputs,
    inputs = list(source_file.files)
    + list(pybin.files),
    executable = list(pybin.files)[0],
    arguments = [f.path for f in source_file.files],
)
```

I get the following output:

```
ERROR: /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/external/assets/BUILD:23:1: Fusing texture from InputFileConfiguredTarget(@assets//:particles/spark_000.bmp) failed: 

Sandboxed execution failed, which may be legitimate (e.g. a compiler error), or due to missing dependencies. To enter the sandbox environment for easier debugging, run the following command in parentheses. On command failure, a bash shell running inside the sandbox will then automatically be spawned:

namespace-sandbox failed: error executing command 
 (cd /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion && \
  exec env - \
  /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion/_bin/namespace-sandbox @/home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion/bazel-sandbox/a41cdbd0-1f9c-41c0-b4f3-61aaeb786afd-1.params -- builders/texture.py external/assets/particles/spark_000.bmp).
src/main/tools/namespace-sandbox.c:771: sandbox root is /home/andreas/.cache/bazel    /_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion/bazel-sandbox/a41cdbd0-1f9c-41c0-b4f3-61aaeb786afd-1
src/main/tools/namespace-sandbox.c:773: working dir is /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion
src/main/tools/namespace-sandbox.c:392: unshare succeeded after 1 tries
src/main/tools/namespace-sandbox.c:558: createdir: /tmp
src/main/tools/namespace-sandbox.c:558: createdir: /nonexistent
src/main/tools/namespace-sandbox.c:573: mount: /bin
src/main/tools/namespace-sandbox.c:573: mount: /etc
src/main/tools/namespace-sandbox.c:573: mount: /lib
src/main/tools/namespace-sandbox.c:573: mount: /lib32
src/main/tools/namespace-sandbox.c:573: mount: /lib64
src/main/tools/namespace-sandbox.c:573: mount: /libx32
src/main/tools/namespace-sandbox.c:573: mount: /run/resolvconf/resolv.conf
src/main/tools/namespace-sandbox.c:573: mount: /sbin
src/main/tools/namespace-sandbox.c:573: mount: /usr/bin
src/main/tools/namespace-sandbox.c:573: mount: /usr/games
src/main/tools/namespace-sandbox.c:573: mount: /usr/include
src/main/tools/namespace-sandbox.c:573: mount: /usr/lib
src/main/tools/namespace-sandbox.c:573: mount: /usr/lib32
src/main/tools/namespace-sandbox.c:573: mount: /usr/libx32
src/main/tools/namespace-sandbox.c:573: mount: /usr/locale
src/main/tools/namespace-sandbox.c:573: mount: /usr/sbin
src/main/tools/namespace-sandbox.c:573: mount: /usr/share
src/main/tools/namespace-sandbox.c:573: mount: /usr/src
src/main/tools/namespace-sandbox.c:573: mount: /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion/bazel-out/local-py3-fastbuild/bin/builders/texture
src/main/tools/namespace-sandbox.c:573: mount: /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion/builders/builder.py
src/main/tools/namespace-sandbox.c:573: mount: /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion/builders/texture.py
src/main/tools/namespace-sandbox.c:573: mount: /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion/external/assets/particles/spark_000.bmp
src/main/tools/namespace-sandbox.c:573: mount: /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion/external/assets/particles/spark_000.bmp.bp
src/main/tools/namespace-sandbox.c:573: mount: /home/andreas/.cache/bazel/_bazel_andreas/install/8850384eb3472503b8c28c60190d6132/_embedded_binaries/build-runfiles
src/main/tools/namespace-sandbox.c:694: arg: builders/texture.py
src/main/tools/namespace-sandbox.c:694: arg: external/assets/particles/spark_000.bmp
src/main/tools/namespace-sandbox.c:707: execvp(argv[0], argv): Permission denied
```

**Comment by dslomov** [comment]
Is builders/texture.py executable? What file is pybin.files[0]?
Try replacing `executable = list(pybin.files)[0],` with `executable = pybin.files_to_run.executable`

**Comment by abergmeier** [comment]
The `Permission denied` was an oversight by me. Referenced the `texture.py` in the label instead of the `pybin`.

Now using `pybin.files_to_run.executable` it fails with:   

    namespace-sandbox failed: error executing command 
      (cd /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion && \
      exec env - \
      /home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion/_bin/namespace-sandbox @/home/andreas/.cache/bazel/_bazel_andreas/4684e913feed063717ca6ee035b0e608/execroot/fusion/bazel-sandbox/9b5579ad-add8-4081-aa19-0136912e5927-0.params -- bazel-out/local-py3-fastbuild/bin/builders/texture external/assets/assets/vfx/generic/vfx_generic_line_004.tga).
    Traceback (most recent call last):
      File "bazel-out/local-py3-fastbuild/bin/builders/texture", line 120, in <module>
    Main()
      File "bazel-out/local-py3-fastbuild/bin/builders/texture", line 75, in Main
    sys.argv[0])
    AssertionError: Cannot find .runfiles directory for bazel-out/local-py3-fastbuild/bin/builders/texture

**Comment by dslomov** [comment]
FYI @philwo @hermione521

**Comment by abergmeier** [comment]
Interesting that I had the `.runfiles` error only once. On my second run, it works!? Is there maybe some dependency not expressed correctly?

**Comment by philwo** [comment]
Could you try this?

```
ctx.action(
    outputs = outputs,
    inputs = list(source_file.files),
    executable = pybin,
    arguments = [f.path for f in source_file.files],
)
```

This should automatically do the right thing and include the files of pybin and setup its runfiles for the action. :)

**Comment by abergmeier** [comment]
This gives me:

    expected file or PathFragment for executable but got RuleConfiguredTarget instead.

**Comment by dslomov** [comment]
Ok, I think @philwo's version should just work (it does not right now, but it should)

**Comment by dslomov** [comment]
(Changed the title accordingly)

**Comment by abergmeier-dsfishlabs** [comment]
@dslomov I think a similar problem is in `repository_ctx.execute`. Should I file a separate issue for that?

**Comment by dslomov** [comment]
Will take a look at this as part of overhauling actions api

**Comment by c-parsons** [comment]
I know this bug is old, so I'm not sure if the use case is still relevant / a blocker. Can anyone confirm?

I don't fully understand why this is required. Can't one access the relevant executable output from [ctx.executable](https://docs.bazel.build/versions/master/skylark/lib/ctx.html#executable) and then pass that as the executable param?

In other words, instead of `executable = pybin` (where `pybin` is presumably a target object) why not `executable = ctx.executable.pybin`

**Comment by laurentlb** [comment]
Closing as obsolete.

You may file a new one (using [the new API](https://docs.bazel.build/versions/master/skylark/lib/ctx.html#actions)) if there's still an issue for you.
